dexios/src/subcommands/decrypt.rs

```diff
@@ -1,263 +1,85 @@
-
-            master_key_decrypted.zeroize();
-            Protected::new(master_key)
-        }
-        HeaderVersion::V5 => {
-            let keyslots = header.keyslots.clone().unwrap();
-            let mut master_key = [0u8; 32];
-            for keyslot in keyslots {
-                let hash_start_time = Instant::now();
-                let key = keyslot
-                    .hash_algorithm
-                    .hash(raw_key.clone(), &keyslot.salt)?;
-                let hash_duration = hash_start_time.elapsed();
-                success!(
-                    "Successfully hashed your key [took {:.2}s]",
-                    hash_duration.as_secs_f32()
-                );
-                let cipher = Ciphers::initialize(key, &header.header_type.algorithm)?;
-
-                let master_key_result =
-                    cipher.decrypt(&keyslot.nonce, keyslot.encrypted_key.as_slice());
-
-                if let Ok(mut master_key_decrypted) = master_key_result {
-                    let len = 32.min(master_key_decrypted.len());
-                    master_key[..len].copy_from_slice(&master_key_decrypted[..len]);
-                    master_key_decrypted.zeroize();
-                    break;
-                }
+    let input_file = stor.read_file(input)?;
+    let header_file = match &params.header_location {
+        HeaderLocation::Embedded => None,
+        HeaderLocation::Detached(path) => {
+            if !overwrite_check(path, params.skip)? {
+                exit(0);
             }
 
-            if master_key != [0u8; 32] {
-                Protected::new(master_key)
-            } else {
-                return Err(anyhow::anyhow!("Unable to find a match with the key you provided (maybe you supplied the wrong key?)"));
-            }
+            Some(stor.read_file(path)?)
         }
     };
 
-    let decrypt_start_time = Instant::now();
+    let raw_key = params.key.get_secret(&PasswordState::Direct)?;
+    let output_file = stor
+        .create_file(output)
+        .or_else(|_| stor.write_file(output))?;
 
-    let streams = DecryptionStreams::initialize(key, &header.nonce, &header.header_type.algorithm)?;
+    if let Some(l) = Arc::get_mut(&mut logger) {
+        l.info(format!("Decrypting {} (this may take a while)", input));
+    }
 
-    streams.decrypt_file(&mut input_file, &mut output_file, &aad)?;
+    // 2. decrypt file
+    let start_time = Instant::now();
+    let mut inner_logger = logger.clone();
+    domain::decrypt::execute(domain::decrypt::Request {
+        header_reader: header_file.as_ref().and_then(|h| h.try_reader().ok()),
+        reader: input_file.try_reader()?,
+        writer: output_file.try_writer()?,
+        raw_key,
+        on_decrypted_header: Some(Box::new(move |header_type| {
+            if let Some(l) = Arc::get_mut(&mut inner_logger) {
+                l.info(format!("Using {} for decryption", header_type.algorithm));
```

Comment:
@brxken128 This is how `cli` can control the progress :) This is just an example, and I didn't like the `paris Logger` implementation, but the gist is this)

Reply:
Everyone has his own area of responsibility

Let logic control logic.
Let the UI control the interface.